### ISSUE TITLE
A bug in abstract_dataset.py

### ISSUE THREAD

**Issue by HaoJia-Alchemist** [SCLBD/DeepfakeBench#29]
You put the shuffle and return code inside the loop. This will result in only the first dataset being returned when using multiple training datasets. It should be placed after the loop ends.

Your code:
<img width="777" alt="image" src="https://github.com/SCLBD/DeepfakeBench/assets/48431661/1c4d7e67-9044-43b8-98b0-6b4a8917a08a">

Right code:
<img width="798" alt="image" src="https://github.com/SCLBD/DeepfakeBench/assets/48431661/8477c702-c23c-4a0b-b505-5dbc13f69563">

**Comment by YZY-stack** [comment]
Please note that we have split multiple dataloaders **outside** this function. See the training script. Thus, it seems not the issue.

**Comment by DIY-Z** [comment]
> Please note that we have split multiple dataloaders **outside** this function. See the training script. Thus, it seems not the issue.

@YZY-stack 

I noticed in the code that it iterates through each item in `config['test_dataset']` only when loading the test dataset. 
https://github.com/SCLBD/DeepfakeBench/blob/1d09eb947d83666b901b3dadbb1852da0c84dcc1/training/train.py#L105-L106
https://github.com/SCLBD/DeepfakeBench/blob/1d09eb947d83666b901b3dadbb1852da0c84dcc1/training/train.py#L86-L89
However, when loading the training dataset, it directly passes the `config` to `pairDataset` or `DeepfakeAbstractBaseDataset`, and so on. When `config['train_dataset']` is set to `[FF-F2F, FF-DF, FF-FS, FF-NT,]`, it only loads the FF-F2F dataset and returns, skipping the rest of the datasets.
https://github.com/SCLBD/DeepfakeBench/blob/1d09eb947d83666b901b3dadbb1852da0c84dcc1/training/train.py#L56-L82
https://github.com/SCLBD/DeepfakeBench/blob/1d09eb947d83666b901b3dadbb1852da0c84dcc1/training/config/detector/ucf.yaml#L20-L21


So I feel it's better to make the code changes suggested by HaoJia-Alchemist.

![](https://user-images.githubusercontent.com/48431661/273435122-8477c702-c23c-4a0b-b505-5dbc13f69563.png)

**Comment by YZY-stack** [comment]
Hello, I'm grateful for the work you guys have done on this code. I'd like to clarify a few points:

Regarding training, it's necessary to combine all the training data (like different elements in a Python list) for an effective training process. In my previous version, the `abstract_dataset` class supported only one training dataset, but I've now fixed this issue in the current version. You can now use multiple datasets for training.

For testing, it's important to evaluate different datasets individually. This requires keeping track of each testing dataset's results separately. That's the reason for the for-loop in the `train.py`, which operates outside the `dataset.py`.

The key change in the `abstract_dataset` code is that now we only pass one dataset name (a string) into the `collect_img_label function`. The logic for collecting training and testing data is set in the `__init__ function`.
